src/server/editorServices.ts

```diff
@@ -1406,47 +1623,101 @@ namespace ts.server {
+                    if (!project) {
+                        project = this.openConfigFile(configFileName, info.fileName);
+                        mapUpdatedProjects.set(configFileName, true);
+                    }
+                    else if (!mapUpdatedProjects.has(configFileName)) {
+                        if (delayReload) {
+                            project.pendingReload = true;
+                            this.delayUpdateProjectGraph(project);
+                        }
+                        else {
+                            this.reloadConfiguredProject(project);
+                        }
+                        mapUpdatedProjects.set(configFileName, true);
+                    }
+                }
+            }
+        }
+
+        /**
+         *  - script info can be never migrate to state - root file in inferred project, this is only a starting point
+         *  - if script info has more that one containing projects - it is not a root file in inferred project because:
+         *    - references in inferred project supercede the root part
+         *    - root/reference in non-inferred project beats root in inferred project
+         */
+        private removeRootOfInferredProjectIfNowPartOfOtherProject(info: ScriptInfo) {
+            if (info.containingProjects.length > 1 &&
+                info.containingProjects[0].projectKind === ProjectKind.Inferred &&
+                info.containingProjects[0].isRoot(info)) {
+                const inferredProject = info.containingProjects[0] as InferredProject;
+                if (inferredProject.isProjectWithSingleRoot()) {
+                    this.removeProject(inferredProject);
+                }
+                else {
+                    inferredProject.removeFile(info);
+                }
             }
-            this.refreshInferredProjects();
         }
 
         /**
          * This function is to update the project structure for every projects.
          * It is called on the premise that all the configured projects are
          * up to date.
          */
-        refreshInferredProjects() {
-            this.logger.info("updating project structure from ...");
+        private refreshInferredProjects() {
+            this.logger.info("refreshInferredProjects: updating project structure from ...");
             this.printProjects();
 
-            const orphantedFiles: ScriptInfo[] = [];
-            // collect all orphanted script infos from open files
             for (const info of this.openFiles) {
+                // collect all orphanted script infos from open files
                 if (info.containingProjects.length === 0) {
-                    orphantedFiles.push(info);
+                    this.assignScriptInfoToInferredProjectIfNecessary(info, /*addToListOfOpenFiles*/ false);
                 }
+                // Or remove the root of inferred project if is referenced in more than one projects

```

Comment:
Based on the `if` condition, this should be "one or more" instead of "more than one"?